File: packages/template-php/src/Parser/BlockTag.php
<?php

declare(strict_types=1);

namespace Polyspec\Template\Parser;

use Polyspec\Template\Ast;
use Polyspec\Template\Expr\Lexer;
use Polyspec\Template\Expr\Parser as ExpressionParser;
use Polyspec\Template\Source;
use Polyspec\Template\TemplateError;

/**
 * Path tokens (GRM-3) and block tag bodies (GRM-13 to GRM-15).
 */
final class BlockTag
{
    private const PATH_CHARS = '/^[A-Za-z0-9_.\/-]$/';
    private const IDENT = '/^[A-Za-z_][A-Za-z0-9_]*$/';

    public function __construct(
        private readonly Source $source,
        private readonly string $template,
        private readonly string $close,
        private readonly int $closeCount,
    ) {
    }

    private function fail(string $code, int $start, int $end, string $message): TemplateError
    {
        return TemplateError::at($code, $this->template, $this->source->lines, $start, $end, $message);
    }

    private function atClose(int $index): bool
    {
        $sequence = str_repeat($this->close, $this->closeCount);

        return substr($this->source->text, $index, strlen($sequence)) === $sequence;
    }

    /**
     * Reads a quoted string or a run of path characters. Returns null at the close delimiter or end of input.
     *
     * @return array{kind: string, value: string, start: int, end: int}|null
     */
    private function readToken(int $index): ?array
    {
        $text = $this->source->text;
        $index = Scanner::skipHorizontalSpace($text, $index);
        if ($index >= strlen($text) || $this->atClose($index)) {
            return null;
        }
        $char = $text[$index];
        if ($char === '"' || $char === "'") {
            [$decoded, $end] = Lexer::stringLiteral($this->source, $index, $this->template);

            return ['kind' => 'path', 'value' => $decoded, 'start' => $index, 'end' => $end];
        }
        $end = $index;
        while ($end < strlen($text) && preg_match(self::PATH_CHARS, $text[$end]) === 1) {
            $end++;
        }
        $value = substr($text, $index, $end - $index);
        if (str_contains($value, '.') || str_contains($value, '/')) {
            return ['kind' => 'path', 'value' => $value, 'start' => $index, 'end' => $end];
        }
        if (preg_match(self::IDENT, $value) === 1) {
            return ['kind' => 'ident', 'value' => $value, 'start' => $index, 'end' => $end];
        }

        return ['kind' => 'ident', 'value' => '', 'start' => $index, 'end' => $index + 1];
    }

    /**
     * GRM-12: `+ path`.
     *
     * @return array{0: string, 1: int} path and the index after it
     */
    public function readIncludePath(int $index): array
    {
        $token = $this->readToken($index);
        if ($token === null || $token['value'] === '') {
            $at = Scanner::skipHorizontalSpace($this->source->text, $index);
            throw $this->fail('E_PARSE_INVALID_PATH', $at, $at + 1, 'include requires a path');
        }
        if ($token['kind'] !== 'path') {
            throw $this->fail('E_PARSE_INVALID_PATH', $token['start'], $token['end'], json_encode($token['value']) . ' is not a path');
        }

        return [$token['value'], $token['end']];
    }

    /**
     * GRM-13 to GRM-15: `# [id] [path] {scope_item}`.
     *
     * @return array{id: string|null, path: string|null, scope: list<array{name: string, expr: array<string, mixed>}>, end: int}
     */
    public function readBlockBody(int $index): array
    {
        $text = $this->source->text;
        $id = null;
        $path = null;
        $scope = [];
        $cursor = $index;
        $token = $this->readToken($cursor);
        if ($token === null || $token['value'] === '') {
            $at = Scanner::skipHorizontalSpace($text, $cursor);
            throw $this->fail('E_PARSE_INVALID_BLOCK_TAG', $at, $at + 1, 'block tag requires an identifier or a path');
        }
        if ($token['kind'] === 'path') {
            $path = $token['value'];
            $cursor = $token['end'];
        } else {
            $id = $token['value'];
            $cursor = $token['end'];
            $token = $this->readToken($cursor);
            if ($token !== null && $token['kind'] === 'path') {
                $path = $token['value'];
                $cursor = $token['end'];
            }
        }
        for (;;) {
            $token = $this->readToken($cursor);
            if ($token === null) {
                break;
            }
            if ($token['kind'] !== 'ident' || $token['value'] === '') {
                throw $this->fail('E_PARSE_INVALID_BLOCK_TAG', $token['start'], $token['end'], 'unexpected ' . json_encode(substr($text, $token['start'], $token['end'] - $token['start'])) . ' in block tag');
            }
            $cursor = $token['end'];
            if (($text[$cursor] ?? '') === ':') {
                $valueStart = $cursor + 1;
                if ($valueStart >= strlen($text) || Scanner::isHorizontalSpace($text[$valueStart])) {
                    throw $this->fail('E_PARSE_INVALID_BLOCK_TAG', $cursor, $cursor + 1, 'scope item requires a value after ":"');
                }
                $parser = new ExpressionParser($this->source, $valueStart, $this->close, $this->closeCount, null, $this->template);
                $scope[] = ['name' => $token['value'], 'expr' => $parser->parsePostfix(true)];
                $cursor = $parser->end();
            } else {
                $scope[] = ['name' => $token['value'], 'expr' => Ast::variable($token['value'], $token['start'], $token['end'])];
            }
        }

        return ['id' => $id, 'path' => $path, 'scope' => $scope, 'end' => $cursor];
    }
}

File: packages/template-rust/src/parser/block_tag.rs
//! Path tokens (GRM-3) and block tag bodies (GRM-13 to GRM-15).

use crate::ast::{Expr, ScopeItem};
use crate::error::{ErrorCode, TemplateError};
use crate::expr::lexer::{LexerOptions, lex_string_literal};
use crate::expr::parser::ExpressionParser;
use crate::parser::scanner::{ident_length, is_horizontal_space, skip_horizontal_space};
use crate::source::Source;

fn is_path_char(byte: u8) -> bool {
    byte.is_ascii_alphanumeric() || matches!(byte, b'_' | b'.' | b'/' | b'-')
}

enum RawToken {
    Path { value: String, start: usize, end: usize },
    Ident { value: String, start: usize, end: usize },
}

/// The parsed body of a block tag.
pub struct BlockBody {
    /// Block identifier.
    pub id: Option<String>,
    /// Template path.
    pub path: Option<String>,
    /// Scope items.
    pub scope: Vec<ScopeItem>,
    /// Byte offset after the body.
    pub end: usize,
}

/// Reader for tag bodies that are not expressions.
pub struct RawTagReader<'a> {
    source: &'a Source,
    close: u8,
    close_count: usize,
}

impl<'a> RawTagReader<'a> {
    /// Creates a reader for a tag with the given close delimiter.
    pub fn new(source: &'a Source, close: u8, close_count: usize) -> RawTagReader<'a> {
        RawTagReader {
            source,
            close,
            close_count,
        }
    }

    fn bytes(&self) -> &'a [u8] {
        self.source.bytes()
    }

    fn at_close(&self, index: usize) -> bool {
        let bytes = self.bytes();
        let sequence: Vec<u8> = std::iter::repeat_n(self.close, self.close_count).collect();
        index < bytes.len() && bytes[index..].starts_with(&sequence)
    }

    fn read_token(&self, index: usize) -> Result<Option<RawToken>, TemplateError> {
        let bytes = self.bytes();
        let index = skip_horizontal_space(bytes, index);
        if index >= bytes.len() || self.at_close(index) {
            return Ok(None);
        }
        let byte = bytes[index];
        if byte == b'"' || byte == b'\'' {
            let (decoded, end) = lex_string_literal(self.source, index)?;
            return Ok(Some(RawToken::Path {
                value: decoded,
                start: index,
                end,
            }));
        }
        let mut end = index;
        while end < bytes.len() && is_path_char(bytes[end]) {
            end += 1;
        }
        let value = self.source.text[index..end].to_string();
        if value.contains('.') || value.contains('/') {
            return Ok(Some(RawToken::Path { value, start: index, end }));
        }
        if !value.is_empty() && ident_length(bytes, index) == value.len() {
            return Ok(Some(RawToken::Ident { value, start: index, end }));
        }
        Ok(Some(RawToken::Ident {
            value: String::new(),
            start: index,
            end: index + 1,
        }))
    }

    /// GRM-12: `+ path`.
    pub fn read_include_path(&self, index: usize) -> Result<(String, usize), TemplateError> {
        match self.read_token(index)? {
            Some(RawToken::Path { value, end, .. }) => Ok((value, end)),
            Some(RawToken::Ident { value, start, end }) if !value.is_empty() => {
                Err(self
                    .source
                    .error(ErrorCode::E_PARSE_INVALID_PATH, start, end, format!("{value:?} is not a path")))
            }
            _ => {
                let at = skip_horizontal_space(self.bytes(), index);
                Err(self
                    .source
                    .error(ErrorCode::E_PARSE_INVALID_PATH, at, at + 1, "include requires a path"))
            }
        }
    }

    /// GRM-13 to GRM-15: `# [id] [path] {scope_item}`.
    pub fn read_block_body(&self, index: usize) -> Result<BlockBody, TemplateError> {
        let bytes = self.bytes();
        let mut id = None;
        let mut path = None;
        let mut scope = Vec::new();
        let mut cursor = index;
        match self.read_token(cursor)? {
            Some(RawToken::Path { value, end, .. }) => {
                path = Some(value);
                cursor = end;
            }
            Some(RawToken::Ident { value, end, .. }) if !value.is_empty() => {
                id = Some(value);
                cursor = end;
                if let Some(RawToken::Path { value, end, .. }) = self.read_token(cursor)? {
                    path = Some(value);
                    cursor = end;
                }
            }
            _ => {
                let at = skip_horizontal_space(bytes, cursor);
                return Err(self.source.error(
                    ErrorCode::E_PARSE_INVALID_BLOCK_TAG,
                    at,
                    at + 1,
                    "block tag requires an identifier or a path",
                ));
            }
        }
        while let Some(token) = self.read_token(cursor)? {
            let (value, start, end) = match token {
                RawToken::Ident { value, start, end } if !value.is_empty() => (value, start, end),
                RawToken::Ident { start, end, .. } | RawToken::Path { start, end, .. } => {
                    let text = self.source.text[start..end.min(self.source.text.len())].to_string();
                    return Err(self.source.error(
                        ErrorCode::E_PARSE_INVALID_BLOCK_TAG,
                        start,
                        end,
                        format!("unexpected {text:?} in block tag"),
                    ));
                }
            };
            cursor = end;
            if bytes.get(cursor) == Some(&b':') {
                let value_start = cursor + 1;
                if value_start >= bytes.len() || is_horizontal_space(bytes[value_start]) {
                    return Err(self.source.error(
                        ErrorCode::E_PARSE_INVALID_BLOCK_TAG,
                        cursor,
                        cursor + 1,
                        "scope item requires a value after \":\"",
                    ));
                }
                let options = LexerOptions {
                    close: Some(self.close),
                    close_count: self.close_count,
                    open_index: None,
                };
                let mut parser = ExpressionParser::new(self.source, value_start, options);
                let expr: Expr = parser.parse_postfix(true)?;
                scope.push(ScopeItem { name: value, expr });
                cursor = parser.end();
            } else {
                scope.push(ScopeItem {
                    name: value.clone(),
                    expr: Expr::Var {
                        name: value,
                        span: [start, end],
                    },
                });
            }
        }
        Ok(BlockBody {
            id,
            path,
            scope,
            end: cursor,
        })
    }
}
